### PR TITLE
SE Channel Attention on Slice Tokens inside PhysicsAttention

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

PR #772 showed that Squeeze-and-Excitation (SE) channel recalibration on **node features** (after the MLP) improved surface MAE. This experiment applies SE at a DIFFERENT location: directly on the **slice tokens** inside PhysicsAttention.

Slice tokens are the core "physics concept" representations in Transolver — they are what the model learns to route nodes into. Applying SE to slice tokens allows the model to selectively suppress or amplify entire "physics modes" (e.g., a slice token tracking boundary layer separation vs. one tracking freestream flow). This is analogous to channel attention on feature maps in vision, but applied at the "prototype" level of the physics decomposition.

**Key distinction from PR #772:** Node SE recalibrates individual node features. Slice token SE recalibrates the physics prototypes that the attention mechanism operates over. These are orthogonal — both can coexist.

**Safety:** Zero-init on the final Linear ensures training starts as identity (no disruption to pretrained behavior).

## Instructions

Add `--se_slice_token` flag. Apply SE recalibration to slice tokens inside `PhysicsAttention.forward`.

### Step 1: Add argument

```python
parser.add_argument('--se_slice_token', action='store_true',
                    help='Apply SE channel attention on slice tokens inside PhysicsAttention')
```

### Step 2: Add SE module to PhysicsAttention.__init__

```python
# In PhysicsAttention.__init__:
if se_slice_token:
    self.se_slice = True
    bottleneck = max(dim_head // 4, 8)  # dim_head=32 typically → bottleneck=8
    self.se_pool = None  # will use mean pooling
    self.se_fc1 = nn.Linear(dim_head, bottleneck)
    self.se_fc2 = nn.Linear(bottleneck, dim_head, bias=False)
    nn.init.zeros_(self.se_fc2.weight)  # Zero-init: starts as identity
else:
    self.se_slice = False
```

Pass `se_slice_token=args.se_slice_token` from the model constructor to each PhysicsAttention layer.

### Step 3: Apply SE after slice token computation

In `PhysicsAttention.forward`, AFTER line 227 (`slice_token = slice_token / ...`) and BEFORE the attention computation (line 232):

```python
# Apply SE recalibration to slice tokens
if self.se_slice:
    # slice_token shape: [bsz, heads, n_slices, dim_head]
    se_in = slice_token.mean(dim=2)  # Pool over slices → [bsz, heads, dim_head]
    se_gate = torch.sigmoid(self.se_fc2(F.gelu(self.se_fc1(se_in))))  # [bsz, heads, dim_head]
    slice_token = slice_token * se_gate.unsqueeze(2)  # [bsz, heads, n_slices, dim_head]
```

### Step 4: Wire through model construction

Where `PhysicsAttention` is instantiated (in TransolverBlock or model config), pass the `se_slice_token` flag.

### Verification

- First epoch: `se_fc2.weight` should be all zeros → `se_gate ≈ 0.5` (sigmoid(0)) → `slice_token *= 0.5`. This is a mild initial scaling that the model can compensate for.
- After training: check if `se_gate` values have differentiated (some channels > 0.5, some < 0.5) — this means the model learned to selectively emphasize/suppress physics modes.
- Log `se_gate.mean()` and `se_gate.std()` at epoch 0 and epoch 100 to verify the mechanism is active.

### Training commands

```bash
cd cfd_tandemfoil && python train.py \
  --agent nezuko --seed 42 \
  --wandb_name "nezuko/se-slice-s42" \
  --wandb_group "se-slice-token-attention" \
  --se_slice_token \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0

# Seed 73: same flags, --seed 73, CUDA_VISIBLE_DEVICES=1,
#   --wandb_name "nezuko/se-slice-s73"
```

## Baseline

| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.742** | < 11.74 |
| p_oodc | 7.643 | < 7.64 |
| **p_tan** | **27.874** | < 27.87 |
| p_re | 6.419 | < 6.42 |

**Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Metrics

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| **p_in** | 11.637 | 12.016 | **11.827** | 11.742 | **+0.7%** ❌ |
| **p_oodc** | 7.557 | 7.685 | **7.621** | 7.643 | **-0.3%** ✅ |
| **p_tan** | 28.885 | 28.612 | **28.748** | 27.874 | **+3.1%** ❌ |
| **p_re** | 6.438 | 6.342 | **6.390** | 6.419 | **-0.5%** ✅ |
| val_loss | 0.2475 | 0.2473 | 0.2474 | — | — |

### Details

- **W&B runs:** [kswkgdcn](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/kswkgdcn) (seed 42, 146 epochs), [hukdqxtc](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/hukdqxtc) (seed 73, 144 epochs)
- **Epochs:** Both ran until wall-clock limit (180 min), reaching ~145 epochs
- **Peak memory:** Standard (no OOM issues on H100 97GB)

### Training commands

```bash
# Seed 42 (GPU 0)
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent nezuko --seed 42 \
  --wandb_name "nezuko/se-slice-s42" \
  --wandb_group "se-slice-token-attention" \
  --se_slice_token \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0

# Seed 73 (GPU 1): same flags, --seed 73, --wandb_name "nezuko/se-slice-s73"
```

### SE Gate Analysis

The SE mechanism is active — `se_fc2.weight` moved away from zero-init:
- **Seed 42:** weight_mean=-0.041, weight_std=0.232, weight_absmax=1.074
- **Seed 73:** weight_mean=-0.038, weight_std=0.272, weight_absmax=1.878

The gate weights differentiated during training (std > 0.2), confirming the SE mechanism learned to selectively modulate slice token channels. The slight negative mean suggests a net tendency to suppress (gate < 0.5) some physics modes.

### What happened

**SE slice token attention did not beat the baseline.** The key concern is p_tan regressing by +3.1% (28.748 vs 27.874 baseline), and p_in regressing by +0.7%. The OOD splits (p_oodc -0.3%, p_re -0.5%) showed marginal improvement but not enough to offset the tandem regression.

The hypothesis was that SE on slice tokens would help the model selectively suppress/amplify "physics modes." While the SE mechanism did learn non-trivial gating (weights differentiated from zero-init), the net effect was harmful — particularly for tandem transfer. This may be because:

1. **Initial scaling disruption:** At epoch 0, se_gate ≈ 0.5 (sigmoid(0)), which halves all slice token magnitudes. Although the model can compensate, this may have pushed learning dynamics in a suboptimal direction during the critical early training phases.
2. **Tandem sensitivity:** Tandem configurations rely on precise inter-foil physics coupling through slice tokens. Adding a global channel gate may disrupt the delicate balance that the existing slice attention mechanism learned.
3. **Orthogonality concern:** While SE on nodes (PR #772) helped because nodes have local, heterogeneous features, slice tokens are already global prototypes — adding another global recalibration on top may be redundant or even destructive.

### Suggested follow-ups

1. **Residual SE gating:** Instead of `slice_token *= gate`, use `slice_token = slice_token + gate * slice_token` with zero-init to start as true identity (gate=0 → no change) rather than sigmoid-based (gate=0.5 → halved).
2. **Per-slice SE (not per-channel):** Gate individual slices rather than channels: pool over dim_head, produce n_slices gates. This would let the model suppress entire physics modes rather than features within modes.
3. **Head-specific SE:** Apply SE separately per attention head rather than shared across heads, to preserve head specialization.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2314 because ADVISOR: Closing. SE channel attention on slice tokens doesn't beat baseline.

**Results:** p_in +0.7%, p_oodc -0.3%, p_tan +3.1%, p_re -0.5%.

The p_tan regression (+3.1%) is the dominant signal. Adding SE recalibration inside PhysicsAttention increases attention expressiveness but hurts tandem generalization — consistent with the pattern we see across all recent attention mechanism modifications (condition tokens, AoA head, SE). The shared K/V attention is already well-tuned regularization; additional recalibration competes with this inductive bias.

The tiny p_oodc and p_re improvements are within noise. Direction closed..
